jocd/src/br/org/certi/jocd/dapaccess/connectioninterface/AndroidUsbDevice.java

```diff
@@ -0,0 +1,467 @@
+        int stringId = applicationInfo.labelRes;
+        this.appName = stringId == 0 ? applicationInfo.nonLocalizedLabel.toString() :
+                context.getString(stringId);
+    }
+
+    /*
+     * Returns all the connected CMSIS-DAP devices.
+     */
+    public List<ConnectionInterface> getAllConnectedDevices() {
+
+        HashMap<String, UsbDevice> usbDevList = usbManager.getDeviceList();
+
+        List<ConnectionInterface> deviceList = new ArrayList<ConnectionInterface>();
+
+        Log.d(TAG, "Listing connected devices...");
+        for(Map.Entry<String, UsbDevice> entry : usbDevList.entrySet()) {
+            UsbDevice device = entry.getValue();
+            Log.d(TAG, "key: " + entry.getKey());
+            Log.d(TAG, "value: " + device);
+
+            AndroidUsbDevice board = new AndroidUsbDevice(context);
+            board.device = device;
+            board.vendorId = device.getVendorId();
+            board.productId = device.getProductId();
+            board.deviceName = device.getDeviceName();
+            board.productName = device.getProductName();
+            board.manufacturerName = device.getManufacturerName();
+            board.serialNumber = device.getSerialNumber();
+
+            // Add this board to the list of devices.
+            deviceList.add(board);
+            Log.d(TAG, "Vendor ID: " + board.vendorId +
+                    "\nProduct ID: " + board.productId +
+                    "\nDevice Name: " + board.deviceName +
+                    "\nProduct Name: " + board.productName +
+                    "\nManufacturer Name: " + board.manufacturerName +
+                    "\nSerial Number: " + board.serialNumber);
+        }
+
+        return deviceList;
+    }
+
+    public void rxHandler() {
+        // TODO
+    }
+
+    /*
+     * Overload to read(timeout).
+     * Use 20ms as the default timeout.
+     */
+    public byte[] read() {
+        return this.read(20);
+    }
+
+    /*
+     * Read data on the IN endpoint associated to the HID interface.
+     */
+    public byte[] read(int timeout) {
+        if (device == null) {
+            Log.e(TAG, "Internal Error. Trying to read from null device");
+            return null;
+        }
+
+        long startTime = System.currentTimeMillis();
+
+        while (rxQueue.isEmpty()) {
+            if (System.currentTimeMillis() - startTime > timeout) {
+                // Timeout.
+                // Read operations should typically take ~1-2ms.
+                // If this exception occurs, then it could indicate
+                // a problem in one of the following areas:
+                // 1. Bad usb driver causing either a dropped read or write
+                // 2. CMSIS-DAP firmware problem cause a dropped read or write
+                // 3. CMSIS-DAP is performing a long operation or is being
+                //    halted in a debugger
+                // TODO Throw and exception.
```

Comment:
Throw the exception was not implemented.

Reply:
This can't be fixed right now because it depends on CmsisDapProtocol. So it won't be fixed on this pull request and it was included on issue #12 for future fix.